File: next.ml
open Printf

type command = 
    | Set of string * string 
    | Get of string 
    | Help 
    | Exit
    | Invalid

let read_cmd (cmd:string): command = 
    let info = String.split_on_char ' ' cmd in match info with
    | ["set"; key; value] -> Set(key, value)
    | ["get"; key] -> Get(key)
    | ["help"] | ["get_all"] -> Help
    | ["exit"] | ["quit"] | ["exit()"] | ["quit()"] | ["q!"] -> Exit
    | _ -> Invalid

let print_help () =
    print_string "this an in memory key value store\n";
    List.iter (
        fun str -> printf "> available command %s\n" str
    ) ["exit|quit|q!"; "help|get_all"; "set"; "get"];
    ()

let rec loop (ht:Hashtbl.t (string, string)) =
    printf ">>> ";
    let cmd = read_cmd (read_line ()) in match cmd with
    | Set(k,v) -> Hashtbl.add ht k v; loop ht
    | Get(k) -> (
        let maybe_value = Hashtbl.find_opt ht k in match maybe_value with 
        | Some(value) -> printf "> %s\n" value
        | None -> ()
    ); loop ht
    | Help -> print_help (); loop ht
    | Invalid -> print_string "invalid cmd\n"; loop ht
    | Exit -> ()

File: next.fs
module Next
open System

let print_help () =
    printfn "This an in memory key value store\n"
    let args = ["exit|quit|q!"; "help|get_all"; "set"; "get"]
    let _ = List.map (function str -> printf "> available command %s\n" str) args
    None

type Command = 
    | Set of key: string * value :string 
    | Get of key: string 
    | Help
    | Invalid

type MaybeAbort = 
    | Continue of cmd: Command 
    | Abort

let getCommand (cmd:string): MaybeAbort = 
    match cmd.Split ' ' with
    | [|"set"; key; value|] -> Continue(Set(key, value))
    | [|"get"; key|] -> Continue(Get(key))
    | [|"help"|] | [|"get_all"|] -> Continue(Help)
    | [|"exit"|] | [|"quit"|] | [|"exit()"|] | [|"quit()"|] | [|"q!"|] -> Abort
    | _ -> Continue(Invalid)

let exeCommand (cmd:Command, dict:Map<string,string>):Option<Map<string,string>> = 
    match cmd with
    | Set(k, v) -> Some (Map.add k v dict)
    | Get(k) -> 
        match Map.tryFind k dict with
        | Some(v) -> printfn "> value: %s" v
        | None -> printfn "> value not found"
        None
    | Invalid -> printfn "beep boop error fatal system"; None
    | Help -> print_help ()

let rec loop (dict:Map<string, string>) =
    printf ">>> ";
    let cmd = getCommand (Console.ReadLine ())
    match cmd with
    | Continue(cmd) ->
        let maybeNewMap = exeCommand(cmd, dict)
        match maybeNewMap with
        | Some(newMap) -> loop newMap
        | None -> loop dict
    | Abort -> ()
